### DISCUSSION TITLE
Re

### DISCUSSION THREAD

**Discussion by zicez** [tinymovr/Tinymovr#241] [Q&A - Tinymovr] [answered]
Hi,
I recently bought a Tinymovr Servo Kit, and I'm having trouble using it. I'm using a USB2CAN adapter from [Innomaker](https://www.inno-maker.com/product/usb2can-cable/). As the device communicates on linux using Socketcan, so I'm running the following command to use Tinymovr Studio and received the following error codes

```
tinymovr --bustype=socketcan --chan=can0
Received id mismatch. Expected: Node: 0x1, Endpoint:0x1a; Got: Node: 0x0, Endpoint:0x24
Received id mismatch. Expected: Node: 0x2, Endpoint:0x1a; Got: Node: 0x0, Endpoint:0x24
Received id mismatch. Expected: Node: 0x3, Endpoint:0x1a; Got: Node: 0x0, Endpoint:0x24
Received id mismatch. Expected: Node: 0x4, Endpoint:0x1a; Got: Node: 0x0, Endpoint:0x24
Received id mismatch. Expected: Node: 0x5, Endpoint:0x1a; Got: Node: 0x0, Endpoint:0x24
Received id mismatch. Expected: Node: 0x6, Endpoint:0x1a; Got: Node: 0x0, Endpoint:0x24
Received id mismatch. Expected: Node: 0x7, Endpoint:0x1a; Got: Node: 0x0, Endpoint:0x24
Received id mismatch. Expected: Node: 0x8, Endpoint:0x1a; Got: Node: 0x0, Endpoint:0x24
Received id mismatch. Expected: Node: 0x9, Endpoint:0x1a; Got: Node: 0x0, Endpoint:0x24
Received id mismatch. Expected: Node: 0xa, Endpoint:0x1a; Got: Node: 0x0, Endpoint:0x24
No Tinymovr instances detected. Exiting shell...
```
I tried to look up the endpoint within the code source, but there is no endpoint for 0x24. Attempting to manually send a remote frame package, it seems to be an ACK Error. The controller board is not responding. I have triple checked the can bus wire connections. The 3v3 testpoint also output the correct voltage value and the white LED is on.

Any suggestions as to how to debug this problem?

**Comment by yconst** [comment]
Hi @zicez , This does not seem like an endpoint being send or received as part of discovery. If there are no other nodes on the bus, I'm wondering if any configuration is needed for your CAN adapter? We have successfully used socketcan with the CANine adapter, in the past without issues.

In addition, if you would like to verify function of the Tinymovr, you could try the UART interface. Please note that the markings on the R5.1 board are wrong, the correct pinouts are described here: https://tinymovr.readthedocs.io/en/latest/hardware/overview.html#connectivity

**Comment by zicez** [comment]
I tested the functionality of my usb to can adapter, and it works correctly. I connected it with another can node on a different computer, and ran the tinymovr script. The other computer was able to receive the remote frame requests successfully (endpoint 0x1a << shifted by 6 for each node id from 1 to 10).

Using the UART interface (using an arduino), I wrote a test to see if the bootloader is on.

```
byte send = 0xAA;
byte receive = 0x7E;
byte inbyte = 0x00;

void setup() {
  // start serial port
  Serial.begin(115200);
  while (!Serial) {
    ;  // wait for serial port to connect. Needed for native USB port only
  }
}

void loop() {
  // if we get a byte
  Serial.write(send);
  if (Serial.available() > 0) {
    // get incoming byte:
    inbyte = Serial.read();
    delay(1000);
  }
  if (inbyte == receive)
  {
      Serial.println("success");
  }
  Serial.println(inbyte, HEX);
  delay(1000);
}
```
I'm not receiving any serial byte back from the board. Is it possible that I have fried the board somehow? I'm using a dewalt 20V battery as input into the tinymovr controller. The LED light is still on and the 3v3 pin is outputting the correct voltage.

**Reply by yconst** [reply]
You are using the same port (presumably UART over USB) to talk to Tinymovr and to the PC. This will not work. Instead, you need to use a different serial port.

In addition, perhaps test with a simpler script, after the unit has booted, in order to test UART function:
(I have not tested the script below, it may contain errors)
```
void setup() {
  // start serial port
  Serial.begin(115200);
  Serial1.begin(115200);
  while (!Serial) {
    ;  // wait for serial port to connect. Needed for native USB port only
  }
  delay(500);
  Serial1.write(".p\n");
  while (Serial1.available() > 0) {
    Serial.print(Serial1.read());
  }
}

void loop() {
  delay(1000);
}
```

**Reply by yconst** [reply]
In addition, please note that Tinymovr's UART port is not 5V tolerant; connecting sth like an Arduino UNO (which is 5V) may damage the board. To connect a 5V board such as an Arduino, a voltage level shifter should be used.

**Reply by zicez** [reply, marked as answer]
You're right. I didn't properly get the Serial port set up correctly. The board is working now. Using an arduino nano 33 to run this.

```
const int bufferSize = 6;
void setup() {
  // start serial port
  pinMode(LED_BUILTIN, OUTPUT);
  digitalWrite(LED_BUILTIN, LOW);
  Serial.begin(9600);
  Serial1.begin(115200);
  while (!Serial) {
    ;  // wait for serial port to connect. Needed for native USB port only
  }
}

void loop() {
  // if we get a byte
  if (Serial.available() > 0) {
      Serial1.write(Serial.read());
  }
  if (Serial1.available() > 0) {
    //Create a place to hold the incoming message
    static char message[bufferSize];
    static unsigned int message_pos = 0;
    while (Serial1.available() > 0){
        //Read the next available byte in the serial receive buffer
        char inByte = Serial1.read();
        //Add the incoming byte to our message
        message[message_pos] = inByte;
        message_pos++;
    }
    for (int i = 0; i < message_pos; i++){
      Serial.print(message[i]);
    }
    Serial.println();
    message_pos = 0;
  }
  delay(100);
}
```

I can command the board through the UART API. Now that I know the board is alive, I will try again with the CAN bus.
